### ISSUE TITLE
[QUESTION]: Hyperparameter Additional Arguments

### ISSUE THREAD

**Issue by spyda90** [chemprop/chemprop#540]
Hello,

First of all thank you for sharing Chemprop with the community.
I am running several hyperparameter optimizations in order to generate several settings to be used in the correspondent training runs.

My question is about additional arguments that can be defined for chemprop_hyperopt as you described in the main page ("_Additional training arguments can also be supplied during submission, and they will be applied to all included training iterations (--epochs, --aggregation, --num_folds, --gpu, --ensemble_size, --seed, etc.")_).

If I set **--ensemble_size** in chemprop_hyperopt, it will be passed automatically in chemprop_train if I set --config_path <hyperopt> in it or should I define again **--ensemble_size** also in the chemprop_train command line?

Obviously, I guess the other additional arguments work as --ensemble_size does.

Thank you in advance!
Francesco

**Comment by cjmcgill** [comment]
Any training arguments that you provide in the command line for hyper parameter optimization will be used in the individual training trials, such as `--ensemble_size 5`. If you provide a config file, all of the arguments in the config file that are relevant for training will be set as arguments in each of the training trials. If you provide an argument in both the command line and in the configuration file, the config file will take precedence. This is true in both training jobs and the training trials within a hyper parameter optimization job.

**Comment by spyda90** [comment]
Thank you very much Chas for your answer.
I am sorry but since I am quite new in using AI-based method I have still doubts. Maybe an example will be useful.

I ran three chemprop_hyperopt commands as follow:
1) chemprop_hyperopt --data_path $Database --dataset_type regression --num_iters 30 --config_save_path ./hyperopt/hyperopt_30iter.json --metric rmse --extra_metrics r2 --seed 0

2) chemprop_hyperopt --data_path $Database --dataset_type regression --num_iters 30 --config_save_path ./hyperopt/hyperopt_30iter_ensemble5.json --metric rmse --extra_metrics r2 --seed 0 **--ensemble_size 5**

3) chemprop_hyperopt --data_path $Database --dataset_type regression --num_iters 30 --config_save_path ./hyperopt/hyperopt_30iter_features.json --metric rmse --extra_metrics r2 --seed 0 **--features_generator rdkit_2d_normalized --no_features_scaling**

If I look at the config file generated (.json) there aren't any label that claims the usage of  ensemble (2) or features (3). 

Thus if I run the same chemprop_train for all the parameters:
chemprop_train --data_path $Database --dataset_type regression --config_path ./hyperopt/$hyperopt --save_dir ./${hyperopt%.*} --metric rmse --extra_metrics r2
Are **--seed, ensemble_size, and --features_generator** going to be used in chemprop_train, if I did not specify them?

I am asking it because I made a bash script where there are mutliple hyperopt command lines bu only one chemprop_train line with no arguments specified.

Thank you in advance!

**Comment by cjmcgill** [comment]
Oh! I misunderstood your question, apologies. I was answering about what would happen if you provided both command line arguments and a config file in your inputs to hyperparameter optimization. 

I think I understand now that you were asking about the config file provided after hyperparameter optimization. This generated config file only covers the parameters over which the search was carried out (selectable with individual keywords if you like). This config file does not include anything else that you provided to hyperparameter optimization in the command line. You will need to explicitly add the ensemble arguments or the featurization arguments to your subsequent training commands, alongside the generated config file.

**Comment by spyda90** [comment]
Dear Chas, thank you! You have now hit the bullseye, thank you!

In terms of best practices, in case I use --ensemble_size 5 in the chemprop_hyperopt should I use the same setting also in the chemprop_train line that is going to use the config_file coming from the former?

Thank you again.

**Comment by cjmcgill** [comment]
In general, you're going to have the best results if you keep the other arguments consistent.

But it will likely be a reasonably good first guess at hyperparameters for small changes in your other arguments, like switching from ensemble 5 to ensemble 10 or similar.

**Comment by spyda90** [comment]
Ok, that is super clear, thank you again!
